src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpWindowsProxy.cs

```diff
@@ -6,45 +6,70 @@
+        private readonly RegistryKey? _internetSettingsRegistry = Registry.CurrentUser?.OpenSubKey("Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings");
+        private MultiProxy _insecureProxy;      // URI of the http system proxy if set
+        private  MultiProxy _secureProxy;       // URI of the https system proxy if set
+        private FailedProxyCache _failedProxies = new FailedProxyCache();
+        private List<string>? _bypass;          // list of domains not to proxy
+        private List<IPAddress>? _localIp;
         private ICredentials? _credentials;
-        private readonly WinInetProxyHelper _proxyHelper;
+        private WinInetProxyHelper _proxyHelper;
         private SafeWinHttpHandle? _sessionHandle;
         private bool _disposed;
+        private EventWaitHandle _waitHandle = new EventWaitHandle(false, EventResetMode.AutoReset);
+        private const int _registrationFlags = Interop.Advapi32.REG_NOTIFY_CHANGE_NAME | Interop.Advapi32.REG_NOTIFY_CHANGE_LAST_SET | Interop.Advapi32.REG_NOTIFY_CHANGE_ATTRIBUTES | Interop.Advapi32.REG_NOTIFY_THREAD_AGNOSTIC;
+        private RegisteredWaitHandle? _registeredWaitHandle;
 
         public static bool TryCreate([NotNullWhen(true)] out IWebProxy? proxy)
         {
-            // This will get basic proxy setting from system using existing
-            // WinInetProxyHelper functions. If no proxy is enabled, it will return null.
-            SafeWinHttpHandle? sessionHandle = null;
-            proxy = null;
+            proxy = new HttpWindowsProxy();
+            return true;
+        }
 
-            WinInetProxyHelper proxyHelper = new WinInetProxyHelper();
-            if (!proxyHelper.ManualSettingsOnly && !proxyHelper.AutoSettingsUsed)
+        public HttpWindowsProxy()
+        {
+
+            if (_internetSettingsRegistry != null)
             {
-                return false;
+                // we register for change notifications so we can react to changes during lifetime.
+                if (Interop.Advapi32.RegNotifyChangeKeyValue(_internetSettingsRegistry.Handle, true, _registrationFlags, _waitHandle.SafeWaitHandle, true) == 0)
+                {
+                    _registeredWaitHandle = ThreadPool.RegisterWaitForSingleObject(_waitHandle, RegistryChangeNotificationCallback, this, -1, false);
+                }
             }
 
+            _proxyHelper = new WinInetProxyHelper();
+            UpdateConfiguration();
+        }
+
+        private static void RegistryChangeNotificationCallback(object? state, bool timedOut)
+        {
+            HttpWindowsProxy proxy = (HttpWindowsProxy)state!;
+            // We need to register for notification every time
+            Interop.Advapi32.RegNotifyChangeKeyValue(proxy._internetSettingsRegistry!.Handle, true, _registrationFlags, proxy._waitHandle.SafeWaitHandle, true);
+            proxy.UpdateConfiguration();
+        }
+
+        private void UpdateConfiguration()
```

Comment:
Is it possible for multiple `UpdateConfiguration()` callbacks to run in parallel if there are multiple registry edits done in sequence?

We may want to ensure that the most recent UpdateConfiguration wins in this case.

Reply:
The notifications do not have content. e.g. I think it would work just fine as long as the updates would finish in sequence they started. 

The problem may be if let say there are part A and B. B get updated and that starts update reading old A. A updates, new config update starts and finishes before first update is done. The first task may be winner writing older configuration. 

If we feel this is concern we have several options. We can simply lock the proxy object in the callback. Since that is only one place the only contention would be registry notification. But I'm not sure if that can cause any troubles for the thread pool. We can also update `ThreadPool.RegisterWaitForSingleObject` to fire only once but then we need new registration every time and I don't know how expensive that is (relative to all the other processing) I was also originally thinking about swapping the order of  `RegNotifyChangeKeyValue` and `UpdateConfiguration` but I think we may miss some notifications in that case.

Reply:
I do not see a problem with taking the lock.